pkg/vault/test/foundry/BufferVaultPrimitive.t.sol

```diff
@@ -508,6 +509,134 @@ contract BufferVaultPrimitiveTest is BaseVaultTest {
+        _addLiquidityToBufferWithRate(
+            bound(firstDepositUnderlying, 0, _userAmount),
+            bound(secondDepositUnderlying, 0, _userAmount),
+            bound(wrappedAmount, 0, _userAmount),
+            bound(rate, 0.0001e18, 1e18)
+        );
+    }
+
+    function _addLiquidityToBufferWithRate(
+        uint256 firstDepositUnderlying,
+        uint256 secondDepositUnderlying,
+        uint256 wrappedAmount,
+        uint256 rate
+    ) internal {
+        // Ensure we're adding more than the minimum, or it will revert.
+        vm.assume(firstDepositUnderlying + wrappedAmount >= MINIMUM_TOTAL_SUPPLY);
+
+        vm.prank(lp);
+        uint256 firstAddLpShares = router.initializeBuffer(waDAI, firstDepositUnderlying, wrappedAmount);
+        assertEq(
+            firstAddLpShares,
+            firstDepositUnderlying + wrappedAmount - MINIMUM_TOTAL_SUPPLY,
+            "Wrong first lpShares added"
+        );
+
+        // Change the rate after initialization.
+        waDAI.mockRate(rate);
+
+        // Predict the amount of shares to receive
+        (uint256 bufferUnderlyingBalance, uint256 bufferWrappedBalance) = vault.getBufferBalance(waDAI);
+        uint256 currentInvariant = bufferUnderlyingBalance + bufferWrappedBalance.mulDown(rate);
+
+        // Shares = current supply (= first shares added) times the invariant ratio.
+        uint256 expectedSecondAddShares = (vault.getBufferTotalShares(waDAI) * secondDepositUnderlying) /
+            currentInvariant;
+
+        // Deposit only underlying the second time.
+        vm.prank(lp);
+        uint256 secondAddLpShares = router.addLiquidityToBuffer(waDAI, secondDepositUnderlying, 0);
+        assertEq(secondAddLpShares, expectedSecondAddShares, "Wrong second lpShares added");
+
+        // stack-too-deep
+        _verifyWithdrawal(secondDepositUnderlying, secondAddLpShares, rate);
+    }
+
+    function _verifyWithdrawal(uint256 secondDepositUnderlying, uint256 secondAddLpShares, uint256 rate) internal {
+        // Predict results of proportional withdrawal of `secondAddLpShares`.
+        uint256 proportionalWithdrawPct = secondAddLpShares.divDown(vault.getBufferTotalShares(waDAI));
+
+        // Get new balances after deposit.
+        (uint256 bufferUnderlyingBalance, uint256 bufferWrappedBalance) = vault.getBufferBalance(waDAI);
+
+        uint256 expectedUnderlyingOut = proportionalWithdrawPct.mulDown(bufferUnderlyingBalance);
+        uint256 expectedWrappedOut = proportionalWithdrawPct.mulDown(bufferWrappedBalance);
+
+        vm.prank(lp);
+        (uint256 removedUnderlying, uint256 removedWrapped) = vault.removeLiquidityFromBuffer(waDAI, secondAddLpShares);
+        assertApproxEqAbs(removedUnderlying, expectedUnderlyingOut, 1e8, "Wrong underlying amount removed");
+        assertApproxEqAbs(removedWrapped, expectedWrappedOut, 1e8, "Wrong wrapped amount removed");
+
+        uint256 totalUnderlyingValue = removedUnderlying + rate.mulDown(removedWrapped);
```

Comment:
I'd call this bufferInvariantAfter, or something with the buffer invariant, to be consistent with the naming we're using. Besides, I'd prefer convertToAssets instead of using the rate directly here.